bin/cml/runner.js

```diff
@@ -245,63 +273,25 @@ const runLocal = async (opts) => {
   proc.stderr.on('data', dataHandler);
   proc.stdout.on('data', dataHandler);
   proc.on('disconnect', () =>
-    shutdown({ ...opts, reason: `runner disconnected` })
-  );
-  proc.on('close', (exit) =>
-    shutdown({ ...opts, reason: `runner closed with exit code ${exit}` })
+    shutdown({ ...opts, error: new Error('runner disconnected') })
   );
+  proc.on('close', (exit) => {
+    const reason = `runner closed with exit code ${exit}`;
+    if (exit === 0) shutdown({ ...opts, reason });
+    else shutdown({ ...opts, error: new Error(reason) });
+  });
 
   RUNNER = proc;
-  ({ id: RUNNER_ID } = await cml.runnerByName({ name }));
-
   if (idleTimeout > 0) {
     const watcher = setInterval(async () => {
-      let idle = RUNNER_JOBS_RUNNING.length === 0;
+      const idle = RUNNER_JOBS_RUNNING.length === 0;
 
       if (RUNNER_TIMER >= idleTimeout) {
-        try {
-          if (cml.driver === 'github') {
-            const job = await cml.runnerJob({ runnerId: RUNNER_ID });
-
-            if (!job && !idle) {
-              winston.error(
-                `Runner is idle as per the GitHub API but busy as per CML internal state. Resetting jobs. Retrying in ${idleTimeout} seconds...`
-              );
-              winston.warn(
-                `CML GitHub driver response: ${JSON.stringify(job)}`
-              );
-              winston.warn(
-                `CML internal state: ${JSON.stringify(RUNNER_JOBS_RUNNING)}`
-              );
-
-              RUNNER_JOBS_RUNNING = [];
-            }
-
-            if (job && idle) {
-              winston.error(
-                `Runner is busy as per the GitHub API but idle as per CML internal state. Retrying in ${idleTimeout} seconds...`
-              );
-
-              idle = false;
-            }
-          }
-        } catch (err) {
-          winston.error(
-            `Error connecting the SCM: ${err.message}. Will try again in ${idleTimeout} secs`
-          );
-
-          idle = false;
-        }
-
-        if (idle) {
-          shutdown({ ...opts, reason: `timeout:${idleTimeout}` });
-          clearInterval(watcher);
-        } else {
-          RUNNER_TIMER = 0;
-        }
+        shutdown({ ...opts, reason: `timeout:${idleTimeout}` });
+        clearInterval(watcher);
       }
 
-      RUNNER_TIMER++;
+      RUNNER_TIMER = idle ? RUNNER_TIMER + 1 : 0;
```

Comment:
Maybe some in code comments for future readers as this seems to a be a hotspot.
```suggestion
       // RUNNER_TIMER only increments if runner is idle
      RUNNER_TIMER = idle ? RUNNER_TIMER + 1 : 0;
```

Reply:
A better comment indeed is `if working we reset the timer`
I consider comments a very bad practice unless is a link to a discussion and not very fan of that either

Reply:
@dacbd

Reply:
Your first pass reading comprehension is better than mine. I'm not suggesting that we comment all over the code.

This particular chunk of logic is a hotspot though and several issues have emanated from the timeout logic and I would prefer it be as explicit as possible even if it adds some lines perhaps this instead:

I have spent numerous hours trying to debug https://github.com/iterative/cml/issues/808 and I'm not convinced the true culprit was ever definitively nailed down

```js
if (idle) {
  RUNNER_TIMER += 1
} else {
  RUNNER_TIMER = 0
}
```

I'll finish with its fine to resolve here and I'm okay with the logic as is, but the next timeout bug I'll be throughing a larger fit about 🙉